[1단계 - 자동차 경주 구현] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
## 안녕하세요!!

우아한테크코스를 통해서 ‘함께 성장하는 동료’가 되고 싶은 ‘디우’라고 합니다.!!
첫 리뷰라서 그런지 많이 떨리기도 하고 얼마나 많은 피드백이 올지 살짝 두렵기도 하면서 한편으론 설레는..
여러 감정이 오가네요ㅎㅎ
귀한 시간 내서 피드백 해주시는 만큼 열심히 배우겠습니다!!

<hr>

이번 미션을 진행하며 페어와 함께 고민하였지만 아직 궁금증으로 남은 몇가지를 질문드리고 싶습니다!!

1. 혹시 랜덤 값과 관련된 메소드를 테스트할 때 좋은 방법이 있을까요?? 
RandomUtil을 다음과 같이 작성하다 보니 특정 값을 Car에 직접 주입해줄 수 있는 방법이 없어서 고민을 하다가..만약 5번을 진행한다고 하면 자동차의 진행이 0보다 크거나 같고 5보다 작거나 같아야 한다는 사실을 통해서 자동차의 진행을 테스트 하였습니다.
```java
    public static boolean checkMoveOrWait() {
        return getRandomNumber() >= CONDITION_VALUE;
    }

    private static int getRandomNumber() {
        return (int) Math.floor(Math.random()*10);
    }
```

```java
    @Test
    @DisplayName("자동차 반복 진행시 정상 동작 확인")
    public void 자동차_반복_진행_테스트() throws Exception {
        //given
        Car car = new Car("woo");
        int count = 5;

        //when
        for (int i = 0; i < count; i++) {
            car.progress();
        }

        //then
        assertThat(car.getPosition())
                .isGreaterThanOrEqualTo(0)
                .isLessThanOrEqualTo(5);
    }
```

그런데 만약 랜덤값을 고정해서 줄 수 있다면 자동차(Car)의 진행(progress)를 확실하게 테스트해볼 수 있을 것 같다는 생각이 듭니다.(정확히 3번 진행시키고 Car의 position값이 3과 같은지)이러한 경우 소스코드 변경을 통해서 테스트를 조금 더 확실히하는 것이 좋을까요?? 혹은 제가 진행한 정도의 테스트면 충분할까요?? 이것도 개발자의 선택이라고 보면 될까요??

2. Setter는 무분별하게 두는 것은 상태 값 변경이 의도하지 않는 곳에서 빈번하게 일어날 수 있기 때문에 좋지 않지만, Getter는 해당 객체의 인스턴스 변수(상태) 값을 변경하지 않기 때문에 적절히 사용해도 된다고 생각해서 RacingGame클래스의 printProgress() 메소드 에서 각 Car객체의 이름과 진행 위치(position)을 얻기위해서 Getter를 사용하였는데요, 이러한 경우에도 Getter 사용을 지양하는게 좋을까요??? 혹은 더 좋은 방법이 있을까요…??(더 좋은 방법이 있다면..힌트를 주시면 스스로 고민해보도록 하겠습니다.)

```java
    private void printProgress() {
        for (Car car : carList) {
            String carName = car.getName();
            int position = car.getPosition();

            OutputView.printProgress(carName, position);
        }
    }
```

<hr>

이외에도 제가 캐치하지 못한 부분이 많다고 생각됩니다. 따금한 피드백 부탁드립니다!
감사합니다!!

tco0427 committed:
test: String 클래스 요구사항1

split 메소드 테스트

tco0427 committed:
refactor: String 클래스 요구사항1 메소드 분리

split 메소드 추출

tco0427 committed:
test : String 요구사항 2

괄호를 제거한 문자열 반환

tco0427 committed:
test: String 요구사항3

String 인덱스 범위 테스트

tco0427 committed:
fix: String 요구사항2

커밋 누락된 부분 확인 후 커밋

tco0427 committed:
test : SetCollection 요구사항 1

Set 크기 확인 테스트

tco0427 committed:
test : SetCollection 요구사항 2

집합 원소 존재유무 테스트

tco0427 committed:
test : SetCollection 요구사항3

집합에 존재하는지 여부에 따라 결과값이 달라지도록 테스트

tco0427 committed:
docs(readme): 문자열 덧셈 계산기 README 작성

기능 요구사항을 토대로 `구현 기능 목록` 작성

tco0427 committed:
docs(readme): 문자열 덧셈 계산기 README 수정

빈 문자열 또는 null 입력인 경우 추가

tco0427 committed:
test : 계산기 요구사항1

빈 문자열 또는 null 입력의 경우 0 반환 여부 테스트 코드 작성.

tco0427 committed:
feat: Calculator 클래스 작성

splitAndSum 메소드 정의하여 `요구사항1` 테스트 코드 통과

tco0427 committed:
test : 계산기 요구사항2

숫자 하나만을 입력하는 경우 해당 숫자 반환 여부 테스트

tco0427 committed:
docs(readme): 문자열 덧셈 계산기 README 작성

숫자 하나만 입력하는 경우 추가 작성

tco0427 committed:
feat : 계산기 요구사항2

숫자 하나만 입력하는 경우 해당 숫자 반환 기능

tco0427 committed:
test: 계산기 요구사항3

숫자 2개를 `,`로 구분하여 합한 후 반환하는지 테스트

tco0427 committed:
feat: 계산기 요구사항3

숫자 2개인 경우 `,`로 구분하여 합하도록 splitAndSum 구현

tco0427 committed:
test: 계산기 요구사항4

구분자 `:` 를 추가하여 구분하는지 테스트

tco0427 committed:
feat : 계산기 요구사항4

구분자 : 추가

tco0427 committed:
test: 계산기 요구사항5

커스텀 구분자(`//` 와 `\n` 사이)를 추가하여 구분하는지 테스트

tco0427 committed:
feat: 계산기 요구사항5

커스텀 구분자(`//` 와 `\n` 사이)로 구분해서 합을 구하도록 splitAndSum 메소드 수정

tco0427 committed:
test : 계산기 요구사항 6

음수를 입력받은 경우 RuntimeException 발생 여부 확인

tco0427 committed:
refactor: 계산기 메소드 추출

getSum()으로 추출가능한 로직 추출하여 코드 가독성을 올린다.

tco0427 committed:
chore: 계산기 메소드 추출

getSum()으로 추출가능한 로직 추출하여 코드 가독성을 올린다.

tco0427 committed:
feat : 계산기 요구사항 6

음수가 입력되는 경우 RuntimeException 발생

tco0427 committed:
feat : 계산기 요구사항 6

음수가 입력되는 경우 RuntimeException 발생

tco0427 committed:
Merge branch 'step1' of https://github.com/tco0427/java-racingcar into step1

tco0427 committed:
refactor: StringSplitter 클래스를 통해서 문자열 분리하는 부분 추출

StringSplitter 클래스를 별도로 두어서 문자열 분리하는 책임을 위임

tco0427 committed:
refactor: StringSplitter 싱글톤 적용

StringSplitter 싱글톤 적용

tco0427 committed:
fix: StringSplitter 클래스 getInstance() 수정

외부에서 getInstance() 메소드 호출을 위해 static으로 수정

tco0427 committed:
refactor : StringSplitter를 통해 분리된 문자열 분리 기능을 활용하도록 리팩토링

StringSplitter를 통해 분리된 문자열 분리 기능을 활용하도록 리팩토링

tco0427 committed:
refactor : 입력값 검증 부분 CalculatorValidation 클래스로 분리

CalculatorValidation에서 입력값을 검증하도록 책임을 분리

tco0427 committed:
refactor : 입력값 검증 부분 CalculatorValidation 클래스로 분리

CalculatorValidation에서 입력값을 검증하도록 책임을 분리

tco0427 committed:
fix : 보다 명확한 싱글톤 패턴 적용을 위하여 final 키워드 사용

tco0427 committed:
fix : 검증 로직 수정

누락 혹은 오류가 발생한 검증 부분 수정

tco0427 committed:
style: 공백 일관성 유지

공백을 일관성 있게 수정하여 코드 스타일 유지

tco0427 committed:
docs(readme): 자동차 경주 미션 README 작성

자동차 경주 미션 '기능 요구사항'을 토대로 '구현 기능 목록' 작성

tco0427 committed:
feat: 사용자 입력 받기

자동차 이름 목록과 시도할 횟수를 입력받는 InputView 구현

tco0427 committed:
feat: Car 객체 생성

자동차 이름, 현재 전진 횟수를 속성으로 가지는 Car 클래스 정의

tco0427 committed:
feat: 자동차 경주 게임 초기화

자동차 경주 게임을 위한 자동차 이름, 전체 시도 횟수 초기 세팅

tco0427 committed:
feat: RandomUtil 클래스 정의

0~9까지 사이의 정수를 생성해 전진 여부를 판단 및 반환

tco0427 committed:
feat: 현재 진행 상황 출력 view 구현

OutputView 에서 '차 이름'과 '전진 위치'를 받아 출력하는 로직 구현

tco0427 committed:
feat: 각 자동차에 대해 게임 진행 및 출력

각 자동차에 대해서 progress()를 호출하여 게임을 진행하고 그 결과를 OutputView를 통해서 출력

tco0427 committed:
feat: 각 자동차에 대해 게임 진행 및 출력

각 자동차에 대해서 progress()를 호출하여 게임을 진행하고 그 결과를 OutputView를 통해서 출력

tco0427 committed:
feat: 시도횟수만큼 게임 진행 반복

시도 횟수만큼 게임 진행을 반복하도록 while()문을 추가

tco0427 committed:
feat: 경주 게임 우승자 선정

경주 게임을 끝내고 난 후 우승자를 선정하도록 구현

tco0427 committed:
fix: 누락된 'Collections' import 추가

tco0427 committed:
feat: 경주 게임 우승자 출력

경주 게임을 끝내고 난 후 우승자를 출력하는 OutputView 로직 추가

tco0427 committed:
feat: Application 구현

main 메소드를 포함하고 있는 Application 클래스 구현

tco0427 committed:
fix: 출력 구문 수정

출력 순서 혹은 출력 구문 수정하였습니다.

tco0427 committed:
fix: 출력 구문 수정

출력 순서 혹은 출력 구문 수정하였습니다.

tco0427 committed:
feat: 자동차 이름 길이 검증

자동차 이름이 1글자 미만 혹은 5글자 초과인 경우 예외 발생

tco0427 committed:
feat: 자동차 이름 길이 검증

자동차 이름이 1글자 미만 혹은 5글자 초과인 경우 예외 발생

tco0427 committed:
feat: 시도 횟수 검증

시도 횟수가 숫자인지 그리고 양수인지를 검증하는 코드 추가

tco0427 committed:
feat: 자동차 이름 공백 검증

자동차 이름에 공백이 들어가면 안되므로 공백이 들어가는지를 추가 검증

tco0427 committed:
feat: 자동차 이름 중복 검증

자동차 이름이 중복되면 안되므로 중복이 있는지 확인

tco0427 committed:
feat: 자동차 이름 중복 검증

자동차 이름이 중복되면 안되므로 중복이 있는지 확인

tco0427 committed:
fix: 에러 메시지를 출력하도록 수정

에러 메시지를 출력할 수 있도록 수정

tco0427 committed:
refactor: 불필요한 toString 제거

Car 클래스의 toString() 메소드가 불필요하므로 제거하였습니다.

tco0427 committed:
fix: Test 에 SuppressWarnings 추가

단위 테스트 메소드에 '한글' 이름을 사용하는 경우 아스키 문자가 아니어서 warning 을 보이므로 `SuppressWarnings` 를 이용하도록 수정

tco0427 committed:
test: 자동차 진행 테스트

자동차 한 번 진행시 '전진' 혹은 '정지' 하는지 테스트

tco0427 committed:
test: 자동차 반복 진행 테스트

자동차 반복 진행시 의도한대로 제대로 동작하는지 확인

tco0427 committed:
test: 자동차 이름 길이 테스트

자동차 이름 길이가 1글자 이상, 5글자 이하인지 확인

tco0427 committed:
test: 자동차 이름 공백 포함 테스트

자동차 이름 공백 포함 여부 테스트

tco0427 committed:
test: 자동차 이름 중복 테스트

자동차 이름이 중복되면 안되므로 이를 테스트

tco0427 committed:
test: 자동차 이름 중복 테스트

자동차 이름이 중복되면 안되므로 이를 테스트

tco0427 committed:
chore: VerificationUtilTest `DisplayName` 어노테이션 포함

tco0427 committed:
test: 시도횟수 검증 테스트

시도횟수가 숫자가 아니면 예외가 제대로 발생하는지, 음수인 경우에도 제대로 예외가 발생하는지 테스트

tco0427 committed:
fix: 누락된 static import 추가

누락된 `validateAttempt` static import 추가

tco0427 committed:
refactor: 매직 넘버를 상수로 치환

매직 넘버에 대해서 상수로 치환해주었습니다.

tco0427 committed:
refactor: 객체에 메시지를 보내도록 수정

Car 객체에서 position 데이터를 꺼내와서 사용하던 방식에서 Car 클래스에 isMaxPosition() 메소드를 두고 메시지를 보내도록 수정해보았습니다.

tco0427 committed:
refactor: 게임을 위한 초기화 부분 분리

자동차 경주 게임을 위한 초기 설정 부분을 별도의 InitUtil로 분리하였습니다.

jjj0611 commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,41 @@
+package racingcar.entity;
```
entity라는 패키지명을 사용하셨네요.
entity는 어떤 의미로 사용하셨을까요?

tco0427 replied:
별다른 의미 없이 entity라는 패키지명을 부여하였는데요,,질문을 받고 고민해보니 domain이라는 이름이 좀 더 맞다고 생각되어 변경하였습니다!

jjj0611 commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,41 @@
+
+import static racingcar.util.RandomUtil.checkMoveOrWait;
+
+public class Car implements Comparable<Car> {
```
자동차의 순위를 판단하는 것이 각각의 자동차에 있어야할까요? 판단자에 있어야 할까요?
판단을 하는 쪽에서 정렬 기준을 넣어주면(오름차순, 내림차순) 그 안에서 순위를 매기는 것이 오름차순으로 해야하는지 내림차순으로 해야하는지가 한 번에 관리될 것 같아요.
하지만 정렬조건이 Car에 있다면 판단을 하는 쪽에서 안으로 들어와서 어떤 기준으로 정렬이 되는지를 알아야 할 것 같아요. 어떻게 생각하시나요?

tco0427 replied:
질문을 듣고 보니 Car보다는 RacingGame쪽에서 판단을 하는 것이 맞는 것 같습니다!! 이를 통해서 향후에 정렬 기준이 바뀌는 것에 Car가 영향을 받지 않을 수 있도록 하였습니다.

jjj0611 commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Car implements Comparable<Car> {
+
+    public static final int INITIAL_VALUE = 0;
```
public으로 선언될 필요가 있을까요?

tco0427 replied:
Car 클래스 내부에서만 사용되는 상수이므로 private이 더 적절한 것 같습니다!

jjj0611 commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,41 @@
+
+    public Car(String name) {
+        this.name = name;
+    }
```
new Car("");
new Car(null);
과 같이 생성되면 어떻게 되나요?
자동차 이름은 비어 있을 수 없다.
자동차 이름은 null일 수 없다.
자동차 이름은 5자 초과일 수 없다.
와 같은 조건은 어디서 체크되어야 할까요?

tco0427 replied:
자동차 생성시 '자동차 이름'에 관한 검증이 필요한데 이는 자동차를 생성하면서 검증하는 것이 맞다고 생각되었습니다. 따라서 이를 Car 객체 생성시 생성자에서 검증할 수 있도록 수정하였습니다!!

jjj0611 commented on src/main/java/racingcar/entity/Car.java:
```diff
@@ -0,0 +1,41 @@
+            return true;
+        }
+        return false;
+    }
```
만약 꼴등을 알려주는 로직이 있다고 할 때,
isWinner(minPosition)이 될텐데 좀 더 범용적인 메서드가 되도록 이름을 변경하면 어떨까요?
인자로 항상 maxPosition이 들어와야 isWinner의 메서드명에 적합한 상태가 되지 않을까요?

tco0427 replied:
Car 클래스에서 maxPosition이 넘어오는지는 알 필요가 없고, 단지 position값이 넘어온다는 것만 알면된다고 생각하였습니다. 또한 이후에 최소값이 winner가 된다고 하였을 때 외부에서 minPosition만 넘겨 동일한 position인지 검사하는데에도 사용할 수 있습니다. 따라서 피드백해주신 것처럼 범용적인 메소드로 활용할 수 있도록 isSamePosition()으로 변경하였습니다!

jjj0611 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.List;
+
+public class OutputView {
```
여기서는 모든 메서드가 static하게 사용되지만, 기본생성자를 막아두지 않으셨네요!
혹시 의도가 있었을까요?

별  다른 의도가 없었다면 코드는 통일감 있게 작성되는 편이 읽는 사람이 더 편하게 읽을 수 있습니다 :)

tco0427 replied:
빠트린 부분이었네요..수정하였습니다!

jjj0611 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public static void printProgress(String carName, int position) {
+        String positionSign = "-".repeat(position);
```
repeat이라는 메서드를 잘 사용해주셨네요!
"-"는 어떤 매직넘버일까요?

자동차 경주게임의 로직을 깊게 알지 않더라도 코드로 읽힐 수 있게 표현해주시면 도움이 될 것 같아요 :)

tco0427 replied:
진행 상태를 나타내는 매직넘버 이므로 PROGRESS_SYMBOL이라는 상수로 수정하였습니다!!

jjj0611 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,23 @@
+
+public class InputView {
+
+    private static final Scanner scanner = new Scanner(System.in);
```
scanner 객체를 잘 사용해주셨네요!
학습차원의 질문으로 static하게 field를 선언해주셨는데요
이렇게 선언할 경우 어떤 일이 일어나나요? 그렇지 않을 때와 어떤 차이가 있나요?

tco0427 replied:
static으로 선언하게 되면 클래스 변수가 되고, 이 경우 생성자를 통해서 객체가 생성되는 시점이 아닌 클래스 로딩 시점에 메모리(메소드 영역)에 올라가게 됩니다. 또한 클래스당 하나가 생성되어 해당 클래스로부터 생성되는 인스턴스 간에 공유하게 된다고 알고 있습니다. 여기서는 InputView의 모든 메소드가 static 이므로 시간적 특성을 고려하여 static 선언을 해주었습니다!!

jjj0611 commented on src/test/java/junitstudy/StringTest.java:
```diff
@@ -0,0 +1,45 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@SuppressWarnings("NonAsciiCharacters")
+public class StringTest {
```
SetCollectionTest와 StringTest 등 학습 테스트 작성을 잘 해주셨네요.
덕분에 해당 애플리케이션에서 어떤 메서드를 주로 사용하셨을지 예상을 할 수 있었습니다 👍

tco0427 replied:
감사합니다!!😀

jjj0611 commented on src/test/java/junitstudy/StringTest.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Test
+    public void 괄호_제거_테스트() throws Exception {
```
throws Exception 시그니처가 필요할까요?
불필요한 시그니처는 삭제해주세요 :)

tco0427 replied:
넵! 수정하였습니다!!

jjj0611 commented on src/test/java/junitstudy/StringTest.java:
```diff
@@ -0,0 +1,45 @@
+    @Test
+    @DisplayName("인덱스 범위 테스트")
+    public void 특정_위치_문자_테스트() throws Exception {
+        assertThatThrownBy(()-> {"abc".charAt(3);})
```
불필요한 괄호는 삭제해주세요!
intellij로 보시면 가이드가 잘 되어 있습니다!

tco0427 replied:
수정하였습니다~_~

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,73 @@
+public class RacingGame {
+
+    public static final int ZERO = 0;
+    public static final int FIRST_INDEX = 0;
```
public하게 선언될 필요가 있을까요?

tco0427 replied:
해당 클래스 내부에서만 사용하는 상수이므로 private이 적절한 것 같습니다!

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,73 @@
+
+    private List<Car> carList = new ArrayList<>();
+    private List<String> winnerNameList = new ArrayList<>();
+    private int totalAttemptCount;
```
3가지 객체가 아래와 같이 선언되었을 때 발생할 수 있는 문제는 어떤게 있을까요?
List<Car>에 따라서 List<String>의 승자를 계속 만들어낼 수 있지 않을까요?
이렇게 선언해두고 사용하시는 것의 이점은 어떤 것이었을까요?

tco0427 replied:
catList를 통해서 구할 수 있는 winnerNameList는 굳이 인스턴스 변수로 둘 필요가 없으므로 제거하였습니다!😀

jjj0611 commented on src/main/java/racingcar/util/VerificationUtil.java:
```diff
@@ -0,0 +1,62 @@
+    public static void validateCarName(String carName) {
+        validateCarNameLength(carName);
+        validateCarNameSpace(carName);
+    }
```
public method로 선언되어 있는데,
VerificationUtil.validateCarName(null)과 같이 호출되면 어떻게 될까요?

tco0427 replied:
NullPointerException이 발생할 수 있다고 생각합니다! 이를 체크해주는 부분을 추가하였습니다.

jjj0611 commented on src/main/java/racingcar/util/VerificationUtil.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    public static void validateDuplication(String[] carNames) {
+        HashSet<String> hashSet = new HashSet<>(Arrays.asList(carNames));
+
+        if (hashSet.size() < carNames.length) {
+            throw new IllegalArgumentException("[ERROR] 자동차 이름이 중복되면 안됩니다.");
+        }
+    }
```
자료구조의 특성을 활용하여 중복 로직 구현을 하셨네요 👍

tco0427 replied:
감사합니다! 👍 👍

jjj0611 commented on src/main/java/racingcar/util/VerificationUtil.java:
```diff
@@ -0,0 +1,62 @@
+        } catch (Exception e) {
+            throw new IllegalArgumentException("[ERROR] 시도 횟수는 숫자여야 합니다.");
+        }
+    }
```
String -> int가 되는 메서드인데,
validateInteger만 보면 검증만을 하는 것으로 보여요!
메서드명을 변경해보면 어떨까요?

tco0427 replied:
좋은 이름은 떠오르지 않았지만...나름 의미를 알 수 있도록 수정해보았습니다.ㅎㅎ

jjj0611 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,34 @@
+
+        if (isOnlyDigit(str)) {
+            return Integer.parseInt(str);
+        }
```
이 부분의 로직이 따로 분기되어야 할 필요가 있을까요?
아래 계산 로직에 녹일 수는 없을까요?

tco0427 replied:
가능하다고 생각합니다! 수정해보았습니다 ㅎㅎ

jjj0611 commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,34 @@
+
+        isValidTokens(tokens);
+
+        return getSum(tokens);
```
valid를 확인하는 과정에서 integer로 바꾸고 있습니다.
그리고 다시 integer로 바꿔서 합계를 하고 있네요.
변환에 대한 중복을 줄일 방법이 있을까요?

tco0427 replied:
피드백을 받고 나서 다시 한 번 생각해보니 String -> int 로 변환하는 로직이 계속해서 중복되고 있었습니다. 이에 대해서 수정을 진행해보았습니다~_~

jjj0611 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,49 @@
+
+        result = Calculator.splitAndSum("");
+        assertThat(result).isEqualTo(0);
+    }
```
1개의 테스트는 1개의 목적을 가지도록 분리하면 어떨까요?

tco0427 replied:
하나의 테스트가 하나의 assert문을 가져 하나의 목적을 위한 단위 테스트가 되도록 수정하였습니다!!

jjj0611 commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,49 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@SuppressWarnings("NonAsciiCharacters")
+public class CalculatorTest {
```
전체적으로 불필요한 throws Exception이라는 시그니처가 사용되고 잇는 것 같아요.
불필요한 내용은 제거해주세요!

tco0427 replied:
넵!! 수정하였습니다~_~

jjj0611 commented on src/test/java/racingcar/entity/CarTest.java:
```diff
@@ -0,0 +1,43 @@
+        assertThat(car.getPosition())
+                .isGreaterThanOrEqualTo(0)
+                .isLessThanOrEqualTo(5);
+    }
```
질문 주신 내용에 대한 답변을 한 번 해보죠.
5번 호출했을 때 0과 5사이의 값이다.

1번 호출했을 때 0과 1사이의 값이다.
2번 호출했을 때 0과 2사이의 값이다.
n번 호출했을 때 0과 n사이의 값이다.

이 모든 문장은 같은 의미를 가집니다.
n에 대한 값을 어떻게 설정할까요?
또 위와 같은 방식이 적절하게 테스트하고 있는 방법일까요?
해당 부분을 잘 테스트하기 위해선 어떻게 해야할까요?

정확히 3번 진행시키고 테스트하면 잘 테스트해볼 수 있겠다는 생각의 방향이 너무 좋아요!
그대로 테스트하기 위해서 활용할 수 있는 것에 대한 고민을 조금만 더 해볼 수 있을까요?
답을 바로 알려드릴 수도 있을 것 같지만, 고민하는 과정이 더 도움이 될 것 같아요!

tco0427 replied:
Car 클래스의 progress 메소드에서 인자로 (랜덤한) 숫자를 받도록 수정함으로써 이를 테스트 해볼 수 있다고 생각하여 수정해보았습니다!

jjj0611 commented on src/test/java/racingcar/util/VerificationUtilTest.java:
```diff
@@ -0,0 +1,71 @@
+import static racingcar.util.VerificationUtil.*;
+
+@SuppressWarnings("NonAsciiCharacters")
+class VerificationUtilTest {
```
불필요한 괄호와 불필요한 시그니처가 여기에도 있네요 :)

tco0427 replied:
수정하였습니다! 😅

jjj0611 commented on src/main/java/calculator/StringSplitter.java:
```diff
@@ -0,0 +1,40 @@
+        return stringSplitter;
+    }
+
+    public String[] parseString(String str) {
```
Integer.parseInt에 착안하여 parseString이라는 이름을 사용하신 것 같아요.
Integer라는 클래스에서 parseInt를 하여 String -> Int가 되는 과정인데,
String[].parse(String)을 해야 조금 더 기존 Java에서 제공하는 것과 동일한 상태가 될 것 같아요.

naming 고민을 조금 더 해보면 어떨까요?

tco0427 replied:
내부에서 수행되는 작업이 customSplit과 defaultSplit이라는 점을 고려하여 splitString()이라는 이름으로 수정해보았습니다..!

jjj0611 commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -0,0 +1,14 @@
+    }
+
+    private static int getRandomNumber() {
+        return (int) Math.floor(Math.random()*10);
```
10에 대한 매직넘버는 요기 그대로 두어도 될까요?!

tco0427 replied:
랜덤값의 범위를 지정해준다는 의미에서 RANDOM_RANGE라는 상수로 치환해보았습니다!

jjj0611 commented on src/main/java/calculator/CalculatorValidation.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.util.regex.Matcher;
+
+public class CalculatorValidation {
```
해당 클래스는 validate를 하는 역할을 하고 있습니다!
Validation과 validator는 어던 의미적 차이가 있을까요?

tco0427 replied:
Validation은 단순히 '검증'이라는 뜻이라면 Validator는 '검증자'라는 의미로 해당 클래스가 validate하는 역할을 한다는 것을 보다 명확히 알 수 있는 것 같다고 생각합니다...

jjj0611 commented on src/main/java/calculator/CalculatorValidation.java:
```diff
@@ -0,0 +1,53 @@
+    private CalculatorValidation() {}
+
+    public static boolean isEmpty(String str) {
+        return (str == null || str.length() == 0);
```
null에 대한 체크를 먼저 해주신 부분이 매우 좋습니다! 👍

tco0427 replied:
감사합니다! 😃

jjj0611 commented on src/main/java/calculator/CalculatorValidation.java:
```diff
@@ -0,0 +1,53 @@
+        for (String token : tokens) {
+            isDigit(token);
+            isNegativeNumber(token);
+        }
```
두 메서드에서 integer로 변환하는 작업을 2번이나 하고 있네요. 그리고 난 후에 바깥에서 한 번 더 integer로 변화하는 작업을 하고 있는데 중복 로직이 너무 많은 것 같아요!

공통된 로직을 뽑아볼까요? :)

jjj0611 commented on src/main/java/racingcar/util/InitUtil.java:
```diff
@@ -0,0 +1,40 @@
+import static racingcar.view.InputView.getAttemptCount;
+import static racingcar.view.InputView.getCarNames;
+
+public class InitUtil {
```
util성 클래스에서 view 단의 메서드를 호출하고 있네요.
클래스의 책임과 역할을 분명히하고, 관계를 잡아보면 어떨까요?

tco0427 replied:
RacingGame에서 carNames와 attempt를 넘겨주도록 하여 Util에서 View의 의존을 끊어주었습니다!

jjj0611 commented on src/main/java/racingcar/util/InitUtil.java:
```diff
@@ -0,0 +1,40 @@
+
+    public static final String SEPARATOR = ",";
+
+    private InitUtil() {
+    }
```
util성 클래스로 static method를 사용할 때 기본 생성자를 막아주는 것은 매우 좋은 습관입니다! 👍
대부분의 util성 클래스에서 잘 막아주셨네요

tco0427 replied:
감사합니다~_~

jjj0611 commented on src/main/java/racingcar/util/InitUtil.java:
```diff
@@ -0,0 +1,40 @@
+
+        validateAttempt(attempt);
+
+        return Integer.parseInt(attempt);
```
validateAttempt에서 Integer.parseInt를 이미 호출했는데,
반환하면서 또 호출하네요. 이 중복 로직들을 제거할 수 있는 좋은 방법이 있을까요?

tco0427 replied:
중복 로직 제거해보았습니다!!ㅎㅎ

jjj0611 reviewed (changes requested):
안녕하세요 디우! 리뷰어 제이입니다.

먼저 제 코드리뷰는 절대값이 아니고, 참고할만한 코멘트 정도로 이해해주시면 될 것 같아요.
따라서 이해가 되지 않는 부분도 있을거고 동의하지 않는 부분도 있을 수 있는데,
같이 이야기 하며 좋은 방향을 찾아가보면 좋을 것 같아요 :)

--- 

질문 주신 것 중 1번은 아래 코멘트에 녹아져 있으니 함께 고민해보시면 좋을 것 같아요.

2번 질문에 대해서는 getter를 사용해야할 때도 분명히 있습니다. getter의 사용을 지양하라는 것은 절대 사용하지 말라는 것이 아니고, 객체의 상태값을 꺼내서 무엇을 하기보다는 상태값을 가지고 있는 객체가 스스로 일하도록 만드는 것의 중요성을 이야기 하는 것입니다. 이런 부분들에 대해서 계속하여 고민해보는 것은 아주 좋은 습관이에요 :)

기능 목록도 잘 작성해주셨고, 테스트도 잘 작성해주셨네요.
커밋 단위도 잘게 쪼개서 커밋을 찍어주셨는데, 중복된 커밋 로그들도 많이 보입니다.
로그를 좀 더 상세히 남기는 방향 혹은 하나의 커밋으로도 묶을 수 있을 것 같아 보입니다 :)

그리고 애플리케이션 로직에서 a,b,,와 같이 이름을 입력하는 것과 a,b,,c와 같이 이름을 입력하는 것에서 차이가 발생하는데, 이 부분은 의도하신대로 작동하는 것이 맞을까요?

위 질문과 더불어서 코멘트 드린 부분 확인해주세요!
만약 코멘트에서 이해되지 않는 부분이 있다면 편하게 말씀해주세요!!

첫 미션 고생하셨어요!

tco0427 committed:
style: 스타일 수정

if 과 조건절 사이의 공백 추가하여 다른 코드 스타일과 통일성 갖추도록 수정하였습니다.

tco0427 committed:
refactor: private 생성자 추가

모든 메소드를 static으로 가지는 유틸성을 가지는 클래스는 객체를 생성할 이유가 없으므로 private 생성자를 만들어 제한

tco0427 committed:
refactor: 정렬하면서 정렬 기준이 오도록 수정

기존 Car 객체 내부에 정렬 기준을 두었던 것에서, 정렬을 하는 쪽에서 정렬 기준을 주어 정렬함으로써 Car 내부로 들어가 정렬 기준을 확인할 필요가 없도록 수정하였습니다. 또한 향후 다른 곳에서 정렬 기준이 바뀌는 것에 영향을 받지 않을 수 있도록 하였습니다.

tco0427 committed:
refactor: 매직넘버에 대해서 private

각 클래스마다 사용하는 매직넘버를 두었습니다. 따라서 해당 매직넘버는 각 클래스에서만 사용하므로 public으로 열 필요가 없다고 판단하였고, private 으로 접근 제어자를 두어 접근 범위를 줄였습니다.

tco0427 committed:
refactor: 자동차 생성 검증 위치 변경

자동차 생성시 '자동차 이름'에 관한 검증(글자 수 혹은 빈 경우 등)이 필요한데 이는 피드백 주신 것과 같이 자동차를 생성하면서 검증하는 것이 맞다고 생각되었습니다. 따라서 Car 생성자에서 이를 검증할 수 있도록 수정하였습니다.

tco0427 committed:
refactor: Car 클래스의 'isWinner' 메소드명 변경

* 'isWinner' -> 'isSamePosition' 으로 메소드명을 변경하였습니다.
* Car 클래스에서 maxPosition이 넘어오는지 알 필요가 없고, 단지 position 값이 넘어오는지만 알면된다고 생각하였다. 또한 이후에 최소값이 winner가 된다고 하였을 때 외부에서 minPosition만 넘겨 동일한 position인지 검사하는데에도 해당 메소드를 활용할 수 있으므로 범용적인 메소드 명으로 변경하였다고 볼 수 있습니다.

tco0427 committed:
refactor: 'OutputView' 클래스 매직 넘버 사용

"-"는 진행 상태를 나타내는 매직넘버이므로 PROGRESS_SYMBOL로, 우승자 출력시 join() 메소드에서 사용하는 ", "는 구분자이므로 이 둘에 대해서 매직넘버를 사용해주었습니다. 이를 통해서 로직을 깊이 알지 않는 사람도 이해가 가능하도록 하였습니다.

tco0427 committed:
refactor: Test코드 'throws Exception' 삭제

Test코드의 `throws Exception` 과 같은 시그니처는 불필요하므로 삭제하였습니다.

tco0427 committed:
refactor: `StringTest` 람다식의 불필요한 {} 삭제

불필요한 람다식의 {}을 삭제해주었습니다.

tco0427 committed:
refactor: `VerificationUtil` 의 validateInteger 메소드명 변경

validateInteger() 메소드 내에서 String -> int 변환 검증을 하며 변환도 수행하므로 translateVerifiedInteger 로 이름을 수정하였습니다.

tco0427 committed:
refactor: CalculatorTest 수정

하나의 테스트가 하나의 assert문을 가지도록, 즉 하나의 목적을 위한 단위 테스트가 되도록 수정하였습니다.

tco0427 committed:
refactor: 불필요한 시그니처 제거

VerificationUtilTest에 포함된 불필요한 시그니처 사용을 제거하여 보다 가독성있는 코드가 되도록 하였습니다.

tco0427 committed:
refactor: 인스턴스 변수의 수를 최소화

carList를 통해서 구할 수 있는 winnerNameList 라는 값을 굳이 인스턴스 변수로 둘 필요가 없으므로 제거

tco0427 committed:
fix: validateCarName에서 null 검사하도록 수정

validateCarName() 메소드가 public으로 열려있기 때문에 넘겨져오는 인자가 null일 수 있다. 따라서 이에 대한 검사를 진행해주어야 NPE이 발생하지 않으므로 코드를 수정하였습니다.

tco0427 committed:
refactor: RandomUtil 의 매직넘버에 상수 적용

랜덤한 값의 범위를 지정하는 10이라는 매직넘버에 대해서 상수로 표현

tco0427 committed:
refactor: Util성 클래스와 view 분리

RacingGame에서 view에 의존하고 Util성 클래스인 InitUtil은 view에 대한 의존을 하지 않도록 수정

tco0427 committed:
refactor: 중복 로직 제거

VerificationUtil에서 검증 이후 값을 반환해주도록 함으로써 Integer.parseInt()가 중복되는 비용을 제거

tco0427 committed:
refactor: 중복 로직 제거

게산기 구현에서 `String -> int` 로의 변환이 계속해서 중복되는 문제를 확인하였고, 이를 최소화 하였습니다.

tco0427 committed:
refactor: CalculatorValidation -> CalculatorValidator 로 이름 수정

Validation은 단순히 '검증'이라는 명사이고, Validator는 '검증자'라는 의미로 해당 클래스가 validate하는 역할을 한다는 것을 보다 명확히 해주었습니다.

tco0427 committed:
style: 상수에 대해서 대문자로 수정

scanner 상수에 대해서 대문자로 수정하였습니다.

tco0427 committed:
refactor: 우승자 구하기 코드에 stream 적용

getWinners()에서 List를 생성하고 getWinnerNameList() 메소드를 별도로 두어 넘기는 등 코드가 복잡하므로 stream을 적용하여 단순화 하였습니다.

tco0427 committed:
refactor: Car 클래스의 progress()에 랜덤값을 넘겨주도록 수정

Car의 진행을 테스트하기 위해서 Car 클래스의 progress()에서 랜덤 값을 매개변수로 받도록 수정하였습니다.

tco0427 committed:
test: 보다 정확히 자동차의 전진 혹은 정지 테스트

수정한 Car 클래스의 progress() 에 따라서 보다 정확하게 자동차의 전진 혹은 정지를 테스트하도록 수정하였습니다.

tco0427 committed:
refactor: 패키지명 변경 entity -> domain

entity 라는 패키지명보다는 domain (해결하고자하는 문제 혹은 관심사)이 보다 적절하므로 수정하였습니다.

tco0427 committed:
style: 누락된 공백 추가

Car 클래스의 progress() 메소드에서 누락된 공백을 추가하여 다른 메소드들과 통일성을 가지도록 하였습니다.

jjj0611 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,36 @@
+public class OutputView {
+
+    private static final String DELIMITER = ", ";
+    public static final String PROGRESS_SYMBOL = "-";
```
접근 제어자를 변경해볼까요??

tco0427 replied:
누락된 부분이네요ㅜ.ㅜ 수정해보도록 하겠습니다!!

jjj0611 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,36 @@
+    public static void printProgress(String carName, int position) {
+        String positionSign = PROGRESS_SYMBOL.repeat(position);
+
+        System.out.println(carName + " : " + positionSign);
```
String.format을 사용해서 조금 더 가독성을 높여보는건 어떨까요??

tco0427 replied:
음... 아래와 같은 형식으로 수정을 말씀해주신 것으로 이해하였습니다...!!
저의 경우에는 String.format을 이용함으로써 코드가 조금 복잡해 보이는데..제가 이해한 것이 맞을까요???ㅜ.ㅜ

```java
System.out.println(String.format("%s : %s", carName, positionSign));
```

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,79 @@
+    private static final int ZERO = 0;
+    private static final int FIRST_INDEX = 0;
+
+    private List<Car> carList = new ArrayList<>();
```
여기서 기본 배열로 한 번 생성하고, 아래쪽에서 재할당을 하고 있네요.
꼭 여기서 한 번 기본생성자로 생성을 해야할까요?

객체의 생성주기를 객체가 필요한 시점으로 설정하면 어떨까요?

tco0427 replied:
말씀해주신대로 initCar() 메소드를 호출하면서 ArrayList가 할당이되므로 생성해줄 필요가 없어보입니다!!
필요시점에 할당해주도록 수정해보도록하겠습니다!! 감사합니다!

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,79 @@
+
+    private void initRacingCarGame() throws IllegalArgumentException {
+        String carNames = getCarNames();
+        carList = initCar(carNames);
```
이 부분이 생성되는 것을 보면 InitUtil에서 생성로직을 담당하고 있죠?
이렇게 어떠한 곳에서 생성을 담당하게 된다는 것은 저 List<Car>로 묶여있는 존재들이
일정한 규칙을 가져야한다는 것을 의미할 수 있어요.
현재는 List<Car>에서 중복되는 자동차 이름이 존재하면 안되는거죠.
즉, List<Car>로 묶여 있는 자료구조가 일종의 로직을 갖게 된다는 것입니다.

2단계 미션을 진행하면서 [일급컬렉션](https://jojoldu.tistory.com/412)에 대해 공부해보고, 적용해보면 좋을 것 같아요!
제가 드린 링크는 참고 정도로만 사용하시고 다른 글들도 많이 참고해보셔도 좋을 것 같아요!

tco0427 replied:
공부한 내용을 정리해보았습니다!
https://velog.io/@tco0427/%EC%9D%BC%EA%B8%89-%EC%BB%AC%EB%A0%89%EC%85%98-Fist-Class-Collection

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,79 @@
+    private static final int FIRST_INDEX = 0;
+
+    private List<Car> carList = new ArrayList<>();
+    private int totalAttemptCount;
```
아래 쪽 List<Car>에 대한 내용과 비슷하게, totalAttemptCount도 물론 int이지만, 기존의 int와는 조금 다른 int죠. 양수만 가능한 int입니다. 

해당 로직을 넣은 객체로 포장해보는건 어떨까요??

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,79 @@
+        initRacingCarGame();
+        startMessage();
+
+        while (totalAttemptCount --> ZERO) {
```
시도횟수의 숫자가 몇이 되어야 끝나는지에 대한 로직을 감쌀 수도 있겠네요!

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    private int getMaxPosition() {
+        Collections.sort(carList, (o1, o2) -> o2.getPosition() - o1.getPosition());
```
list의 sort 메서드가 아닌 Collections.sort를 사용한 이유가 있을까요??
(변경하라고 말씀드리는 것은 아닙니다! 특별한 이유가 있었나 궁금했어요 😄)

tco0427 replied:
오! list의 sort() 메소드를 활용할 생각을 하질 못했습니다...(제가 아직 부족한 부분이 많네요..)
Collections의 sort() 내부에서 list의 sort()를 호출하는 것을 보니 성능에는 차이가 없어보이고, list의 sort()를 사용하는 것이 코드가 다음과 같아질 수 있어 보다 가독성 있는 코드인 것 같아 변경하였습니다!!

```java
carList.sort((o1, o2) -> o2.getPosition() - o1.getPosition());
```

jjj0611 commented on src/main/java/racingcar/game/RacingGame.java:
```diff
@@ -0,0 +1,79 @@
+    private List<String> getWinners() {
+        int maxPosition = getMaxPosition();
+
+        return carList.stream()
+                .filter(car -> car.isSamePosition(maxPosition))
+                .map(car -> car.getName())
+                .collect(Collectors.toList());
```
Stream 사용법에 대해서 조금씩 익숙해져가는 것 같네요!
강력한 도구니 잘 숙지해두시면 좋을 것 같아요! 🙂

tco0427 replied:
정말 편리한 도구인거 같습니다🙂🙂

jjj0611 commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -0,0 +1,13 @@
+package racingcar.util;
+
+public class RandomUtil {
+
+    public static final int RANDOM_RANGE= 10;
+
+    private RandomUtil() {
+    }
+
+    public static int getRandomNumber() {
+        return (int) Math.floor(Math.random()*RANDOM_RANGE);
+    }
+}
```
전체적으로 formatting이 안맞는 것 같아요!
intellij의 cmd + option + l을 누르면 자동 포멧팅이 됩니다!

tco0427 replied:
오!! 좋은 기능 감사합니다~_~

jjj0611 commented on src/main/java/racingcar/util/VerificationUtil.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Arrays;
+import java.util.HashSet;
+
+public class VerificationUtil {
```
이러한 VerificationUtil을 따로 만들었을 때 어떤 장점과 단점이 존재할까요?

tco0427 replied:
검증에 대한 책임을 유틸성 클래스에 위임할 수 있다는 것이 장점이라고 생각합니다! 따라서 요구사항(글자가 10글자까지 허용되도록)이 변경되어도 해당 클래스에서만 변경이 일어날 것이라고 생각합니다! 하지만 해당 검증이 필요하지 않은 도메인에서도 사용 가능하다는게 단점으로 작용할 수도 있다고 생각합니다..!! (네이밍이 명확하지 않은 경우 다른 검증 메소드를 불러 호출할 수도 있습니다!)

혹시 제가 생각하지 못한...또다른 장점과 단점을 알 수 있을까요...?? ㅜㅜ 막상 이런 질문이 들어오니 제대로된 답변이 어렵네요..ㅎㅎ

jjj0611 reviewed (approved):
안녕하세요 디우! 전체적으로 잘 구현해주셨네요!
1단계는 여기서 머지하고, 코멘트 드린 내용은 다음단계에서 반영해주세요!

자동차의 진행에 대해서는 테스트를 잘 진행해주셨는데,
자동차들에 대한 테스트가 조금 빈약한 것 같아요.
각 자동차에서 1등을 추출하는 것도 하나의 기능이고, 로직일 것 같은데 어떻게 테스트할 수 있을까요?
스스로 부족하다고 생각하는 테스트에 대해서도 조금 더 보강해서 2단계에 녹여주시면 좋을 것 같아요 :)

1단계는 여기서 머지하겠습니다!
